src/mgr/ActivePyModules.cc

```diff
@@ -608,51 +608,15 @@ PyObject* ActivePyModules::get_counter_python(
     const std::string &svc_id,
     const std::string &path)
 {
-  PyThreadState *tstate = PyEval_SaveThread();
-  Mutex::Locker l(lock);
-  PyEval_RestoreThread(tstate);
+  return with_perf_counters(extract_counters, svc_name, svc_id, path);
+}
 
-  PyFormatter f;
-  f.open_array_section(path.c_str());
-
-  auto metadata = daemon_state.get(DaemonKey(svc_name, svc_id));
-  if (metadata) {
-    Mutex::Locker l2(metadata->lock);
-    if (metadata->perf_counters.instances.count(path)) {
-      auto counter_instance = metadata->perf_counters.instances.at(path);
-      auto counter_type = metadata->perf_counters.types.at(path);
-      if (counter_type.type & PERFCOUNTER_LONGRUNAVG) {
-        const auto &avg_data = counter_instance.get_data_avg();
-        for (const auto &datapoint : avg_data) {
-          f.open_array_section("datapoint");
-          f.dump_unsigned("t", datapoint.t.sec());
-          f.dump_unsigned("s", datapoint.s);
-          f.dump_unsigned("c", datapoint.c);
-          f.close_section();
-        }
-      } else {
-        const auto &data = counter_instance.get_data();
-        for (const auto &datapoint : data) {
-          f.open_array_section("datapoint");
-          f.dump_unsigned("t", datapoint.t.sec());
-          f.dump_unsigned("v", datapoint.v);
-          f.close_section();
-        }
-      }
-    } else {
-      dout(4) << "Missing counter: '" << path << "' ("
-              << svc_name << "." << svc_id << ")" << dendl;
-      dout(20) << "Paths are:" << dendl;
-      for (const auto &i : metadata->perf_counters.instances) {
-        dout(20) << i.first << dendl;
-      }
-    }
-  } else {
-    dout(4) << "No daemon state for "
-              << svc_name << "." << svc_id << ")" << dendl;
-  }
-  f.close_section();
-  return f.get();
+PyObject* ActivePyModules::get_latest_counter_python(
+    const std::string &svc_name,
+    const std::string &svc_id,
+    const std::string &path)
+{
+  return with_perf_counters(extract_latest_counter, svc_name, svc_id, path);
```

Comment:
Since extract_latest_counter is only used here, I'd move the definition inline here as a lambda

Reply:
(same for extract_counters)